in_toto/verifylib.go

```diff
@@ -0,0 +1,517 @@
+            }
+
+            // Normalize optional source and destination prefixes, i.e.
+            // if there is a prefix, then add a trailing slash if not there yet
+            for _, prefix := range []string{"srcPrefix", "dstPrefix"} {
+              if ruleData[prefix] != "" &&
+                  ! strings.HasSuffix(ruleData[prefix], "/") {
+                ruleData[prefix] += "/"
+              }
+            }
+            // Iterate over queue and add consumed artifacts
+            // consumed is subtracted from queue
+            var consumed []string
+            for _, srcPath := range queue {
+              // Remove optional source prefix from source artifact path
+              // Noop if prefix is empty, or artifact does not have it
+              srcBasePath := strings.TrimPrefix(srcPath, ruleData["srcPrefix"])
+
+              // Ignore artifacts not matched by rule pattern
+              matched, err := filepath.Match(ruleData["pattern"], srcBasePath)
+              if err != nil || !matched {
+                continue
+              }
+
+              // Construct corresponding destination artifact path, i.e.
+              // an optional destination prefix plus the source base path
+              dstPath := osPath.Join(ruleData["dstPrefix"], srcBasePath)
+
+              // Try to find the corresponding destination artifact
+              dstArtifact, exists := dstArtifacts[dstPath]
+              // Ignore artifacts without corresponding destination artifact
+              if !exists {
+                continue
+              }
+
+              // Ignore artifact pairs with no matching hashes
+              if !reflect.DeepEqual(artifacts[srcPath], dstArtifact) {
+                continue
+              }
+
+              // Only if a source and destination artifact pair was found and
+              // their hashes are equal, will we mark the source artifact as
+              // successfully consumed, i.e. it will be removed from the queue
+              consumed = append(consumed, srcPath)
+            }
+            queue = Subtract(queue, consumed)
+
+          case "allow":
+            consumed := FnFilter(ruleData["pattern"], queue)
+            queue = Subtract(queue, consumed)
+
+          case "disallow":
+            disallowed := FnFilter(ruleData["pattern"], queue)
+            if len(disallowed) > 0 {
+              return fmt.Errorf("Artifact verification failed for %s '%s'." +
+                  " Artifact(s) %s disallowed by rule %s.",
+                  reflect.TypeOf(itemI), itemName, disallowed, rule)
+            }
+
+          // TODO: Support create, modify, delete rules
+          default:
+              return fmt.Errorf("Cannot process artifact rule '%s'. We" +
+                  " don't support rules of type '%s'", rule,
+                  strings.ToUpper(ruleData["type"]))
+        }
+      }
+    }
+  }
+  return nil
+}
+
+
+func ReduceStepsMetadata(layout Layout,
```

Comment:
What does Reducing steps metadata actually entail?

Reply:
It takes multiple links corresponding to a given step of the layout and merges them into a single link to make further processing (artifact rules) easier. If the links have different materials or different products it fails.

IIRC you, @RohanBhirangi and I tried hard to find a proper name for this function, when Rohan added it to the reference implementation, and this was the best compromise we could come up with.

I am still open for alternative names.